Source/Game.cs

```diff
@@ -135,215 +135,235 @@ public void Goto(Transition next)
-				    if (next != last)
-				    {
-					    Music.Stop();
-					    Music = Audio.Play(next);
-					    if (Music)
-						    Music.SetCallback(audioEventCallback);
-				    }
-			    }
-
-			    // switch ambience
-			    {
-				    var last = Ambience.IsPlaying && lastScene != null ? lastScene.Ambience : string.Empty;
-				    var next = nextScene?.Ambience ?? string.Empty;
-				    if (next != last)
-				    {
-					    Ambience.Stop();
-					    Ambience = Audio.Play(next);
-				    }
-			    }
-
-			    // in case new music was played
-			    Save.Instance.SyncSettings();
-			    transitionStep = TransitionStep.FadeIn;
-		    }
-		    else if (transitionStep == TransitionStep.FadeIn)
-		    {
-			    if (transition.ToBlack == null || transition.ToBlack.IsFinished)
-			    {
-				    transitionStep = TransitionStep.None;
-				    transition = new();
-			    }
-			    else
-			    {
-				    transition.ToBlack.Update();
-			    }
-		    }
-		    else if (transitionStep == TransitionStep.None)
-		    {
-			    // handle audio beat events on main thread
-			    if (audioBeatCounterEvent)
-			    {
-				    audioBeatCounterEvent = false;
-				    audioBeatCounter++;
-
-				    if (scene is World world)
-				    {
-					    foreach (var listener in world.All<IListenToAudioCallback>())
-						    (listener as IListenToAudioCallback)?.AudioCallbackEvent(audioBeatCounter);
-				    }
-			    }
-		    }
-
-
-		    if (scene is not Celeste64.Startup)
-		    {
-			    // toggle fullsrceen
-			    if ((Input.Keyboard.Alt && Input.Keyboard.Pressed(Keys.Enter)) || Input.Keyboard.Pressed(Keys.F4))
-				    Save.Instance.ToggleFullscreen();
-
-			    // reload state
-			    if (Input.Keyboard.Ctrl && Input.Keyboard.Pressed(Keys.R) && !IsMidTransition)
-			    {
-				    if (scene is World world)
-				    {
-					    Goto(new Transition()
-					    {
-						    Mode = Transition.Modes.Replace,
-						    Scene = () => new World(world.Entry),
-						    ToPause = true,
-						    ToBlack = new AngledWipe(),
-						    PerformAssetReload = true
-					    });
-				    }
-				    else
-				    {
-					    Goto(new Transition()
-					    {
-						    Mode = Transition.Modes.Replace,
-						    Scene = () => new Titlescreen(),
-						    ToPause = true,
-						    ToBlack = new AngledWipe(),
-						    PerformAssetReload = true
-					    });
-				    }
-			    }
-		    }
-        }
+		if (TASControls.ToggleInfoGUI.Pressed)
+		{
+			imGuiEnabled = !imGuiEnabled;
+		}
+
+		if (imGuiEnabled)
+		{
+			imGuiRenderer.Update();
+		}
+
+		int loops = Manager.FrameLoops; // Copy to local variable, so it doesn't update while iterating
+		for (int i = 0; i < loops; i++)
+		{
+			// We need to manually update the input on the 2nd iteration and forward
+			if (i >= 1) m_Input_Step.Invoke(null, []);
+
+			TASMod.Update();
+			if (Manager.IsPaused())
+			{
+				scenes.TryPeek(out var camscene);
+				if (!(camscene is World)) continue;

```

Comment:
This can be turned in `if (!scenes.TryPeek(our var camscene) || camscene is not World world) continue;`